Describe string cast to JSON handling (#10463)
trinodb/trino: docs/src/main/sphinx/functions/json.rst

Cast to JSON
------------

The following types can be cast to JSON:

* ``BOOLEAN``
* ``TINYINT``
* ``SMALLINT``
* ``INTEGER``
* ``BIGINT``
* ``REAL``
* ``DOUBLE``
* ``VARCHAR``

Additionally, ``ARRAY``, ``MAP``, and ``ROW`` types can be cast to JSON when
the following requirements are met:

* ``ARRAY`` types can be cast when the element type of the array is one

>     of the supported types.

Reviewer: ```suggestion
    of the supported types listed above.
```

Author: Google style guide advises against referring to other sections of a doc with directions.

> * ``MAP`` types can be cast when the key type of the map is ``VARCHAR`` and
>     the value type of the map a supported type,
> * ``ROW`` types can be cast when every field type of the row is a supported
>     type.

Reviewer: ```suggestion
* ``MAP`` types can be cast when the key of the map is of type ``VARCHAR`` and
    the value of the map is one of the supported types listed above.
* ``ROW`` types can be cast when every field of the row is one of the supported
    types listed above.
```

Author: Same point as above

.. note::

    Cast operations with supported :ref:`character string types
    <string-data-types>` treat the input as a string, not validated as JSON.
    This means that a cast operation with a string-type input of invalid JSON
    results in a succesful cast to invalid JSON.

>     Instead, consider using the :ref:`json_parse <json-parse>` function to

Reviewer: ```suggestion
    Instead, consider using the :func:`json_parse` function to
```

Does this render and link correctly? No need to add anchors.

Author: It does, good catch.

------------

Casting from ``BOOLEAN``, ``TINYINT``, ``SMALLINT``, ``INTEGER``,
``BIGINT``, ``REAL``, ``DOUBLE``, or ``VARCHAR`` to JSON is supported.

Casting from ``ARRAY``, ``MAP``, or ``ROW`` is supported when the element type of
the array is one of the supported types, when the key type of the map
is ``VARCHAR`` and value type of the map is one of the supported types,

> or when every field type of the row is one of the supported types.

Reviewer: Would read better as a bulleted list IMO.

Author: Fixed, and cleaned up the following paragraph similarly

.. note::

    Cast operations with supported:ref:`character string types
    <string-data-types>` treat the input as a string, not validated as JSON.
    This means that a cast operation with a string-type input of invalid JSON

>     results in a succesful cast to invalid JSON.

Reviewer: Seems like a bug we should be able to fix. Especially since other CASTs (e.g. string -> number) fail with such invalid inputs and hence the JSON cast behaviour is inconsistent with others. cc: @martint WDYT?

BTW it would help to mention some alternative here if the user wants to "create" JSON. e.g. `json_parse`.

Reviewer 2: @hashhar not a bug (i mean, hm, ... confusing design), but yes, we should link to `json_parse` here.

Reviewer 3: If I recall correctly, this is just a performance optimization. The cast is effectively a no op -- the representation of a JSON type is still a string internally. If the case does validation, then any operation on the resulting JSON value (json_extract, etc) will have to parse the double twice: once for the cast and once for accessing some part of the value.
